src/lib.rs

```diff
@@ -3647,4 +3647,131 @@ export function a(a: A): B {
       assert!(iterator.next().is_none());
     }
   }
+
+  #[tokio::test]
+  async fn test_resolver_execution_and_types_resolution() {
+    #[derive(Debug)]
+    struct ExtResolver;
+
+    impl crate::source::Resolver for ExtResolver {
+      fn resolve(
+        &self,
+        specifier_text: &str,
+        referrer: &ModuleSpecifier,
+        mode: ResolutionMode,
+      ) -> Result<ModuleSpecifier, crate::source::ResolveError> {
+        let specifier_text = match mode {
+          ResolutionMode::Types => format!("{}.d.ts", specifier_text),
+          ResolutionMode::Execution => format!("{}.js", specifier_text),
+        };
+        Ok(resolve_import(&specifier_text, referrer)?)
+      }
+    }
+
+    let mut loader = setup(
+      vec![
+        (
+          "file:///a/test01.ts",
+          Source::Module {
+            specifier: "file:///a/test01.ts",
+            maybe_headers: None,
+            content: r#"
+            import a from "./a";
```

Comment:
Notice how the specifier is `"./a"` here, but it resolves to a `.js` files for the code resolution and a `.d.ts` file for the types resolution because of how the resolver is implemented.